File: bleep-test-runner/src/main/java/bleep/testing/runner/TestProtocol.java
package bleep.testing.runner;

import java.util.ArrayList;
import java.util.List;

/**
 * Simple JSON-based protocol for communication between bleep and forked test JVMs.
 *
 * <p>Outputs valid JSON that can be parsed by circe on the Scala side. Each message is a single
 * line of JSON.
 */
public final class TestProtocol {

  private TestProtocol() {}

  // === Commands (parent -> forked JVM) ===

  public static final String CMD_RUN_SUITE = "RunSuite";
  public static final String CMD_SHUTDOWN = "Shutdown";
  public static final String CMD_GET_THREAD_DUMP = "GetThreadDump";

  // === Responses (forked JVM -> parent) ===

  public static final String RESP_READY = "Ready";
  public static final String RESP_TEST_STARTED = "TestStarted";
  public static final String RESP_TEST_FINISHED = "TestFinished";
  public static final String RESP_SUITE_DONE = "SuiteDone";
  public static final String RESP_LOG = "Log";
  public static final String RESP_ERROR = "Error";
  public static final String RESP_THREAD_DUMP = "ThreadDump";

  // === Response encoding (forked JVM outputs these) ===

  public static String encodeReady() {
    return "{\"type\":\"Ready\"}";
  }

  public static String encodeTestStarted(String suite, String test) {
    return "{\"type\":\"TestStarted\",\"data\":{\"suite\":"
        + jsonString(suite)
        + ",\"test\":"
        + jsonString(test)
        + "}}";
  }

  public static String encodeTestFinished(
      String suite, String test, String status, long durationMs, String message, String throwable) {
    StringBuilder sb = new StringBuilder();
    sb.append("{\"type\":\"TestFinished\",\"data\":{");
    sb.append("\"suite\":").append(jsonString(suite));
    sb.append(",\"test\":").append(jsonString(test));
    sb.append(",\"status\":").append(jsonString(status));
    sb.append(",\"durationMs\":").append(durationMs);
    if (message != null) {
      sb.append(",\"message\":").append(jsonString(message));
    }
    if (throwable != null) {
      sb.append(",\"throwable\":").append(jsonString(throwable));
    }
    sb.append("}}");
    return sb.toString();
  }

  public static String encodeSuiteDone(
      String suite, int passed, int failed, int skipped, int ignored, long durationMs) {
    return "{\"type\":\"SuiteDone\",\"data\":{"
        + "\"suite\":"
        + jsonString(suite)
        + ",\"passed\":"
        + passed
        + ",\"failed\":"
        + failed
        + ",\"skipped\":"
        + skipped
        + ",\"ignored\":"
        + ignored
        + ",\"durationMs\":"
        + durationMs
        + "}}";
  }

  public static String encodeLog(String level, String message) {
    return "{\"type\":\"Log\",\"data\":{\"level\":"
        + jsonString(level)
        + ",\"message\":"
        + jsonString(message)
        + "}}";
  }

  public static String encodeError(String message, String throwable) {
    StringBuilder sb = new StringBuilder();
    sb.append("{\"type\":\"Error\",\"data\":{");
    sb.append("\"message\":").append(jsonString(message));
    if (throwable != null) {
      sb.append(",\"throwable\":").append(jsonString(throwable));
    }
    sb.append("}}");
    return sb.toString();
  }

  /** Encode thread dump information as a list of threads with their stack traces */
  public static String encodeThreadDump(List<ThreadDumpEntry> threads) {
    StringBuilder sb = new StringBuilder();
    sb.append("{\"type\":\"ThreadDump\",\"data\":{\"threads\":[");
    boolean first = true;
    for (ThreadDumpEntry entry : threads) {
      if (!first) sb.append(",");
      first = false;
      sb.append("{\"name\":").append(jsonString(entry.name));
      sb.append(",\"state\":").append(jsonString(entry.state));
      sb.append(",\"stackTrace\":[");
      boolean firstLine = true;
      for (String line : entry.stackTrace) {
        if (!firstLine) sb.append(",");
        firstLine = false;
        sb.append(jsonString(line));
      }
      sb.append("]}");
    }
    sb.append("]}}");
    return sb.toString();
  }

  /** Thread dump entry for encoding */
  public static class ThreadDumpEntry {
    public final String name;
    public final String state;
    public final List<String> stackTrace;

    public ThreadDumpEntry(String name, String state, List<String> stackTrace) {
      this.name = name;
      this.state = state;
      this.stackTrace = stackTrace;
    }
  }

  // === Command parsing (forked JVM receives these) ===

  /**
   * Parse a command from a JSON line. Format:
   * {"type":"RunSuite","data":{"className":"...","framework":"...","args":[...]}} or
   * {"type":"Shutdown"}
   */
  public static ParsedCommand parseCommand(String line) {
    if (line == null || line.isEmpty()) {
      return new ParsedCommand.Invalid("Empty command");
    }

    try {
      // Find the type field
      String type = extractStringField(line, "type");
      if (type == null) {
        return new ParsedCommand.Invalid("Missing type field");
      }

      if (CMD_SHUTDOWN.equals(type)) {
        return new ParsedCommand.Shutdown();
      } else if (CMD_GET_THREAD_DUMP.equals(type)) {
        return new ParsedCommand.GetThreadDump();
      } else if (CMD_RUN_SUITE.equals(type)) {
        // Extract data object fields
        int dataStart = line.indexOf("\"data\"");
        if (dataStart < 0) {
          return new ParsedCommand.Invalid("Missing data field for RunSuite");
        }
        String dataSection = line.substring(dataStart);

        String className = extractStringField(dataSection, "className");
        String framework = extractStringField(dataSection, "framework");
        List<String> args = extractStringArray(dataSection, "args");

        if (className == null || framework == null) {
          return new ParsedCommand.Invalid("Missing className or framework");
        }

        return new ParsedCommand.RunSuite(
            className, framework, args != null ? args : new ArrayList<String>());
      } else {
        return new ParsedCommand.Invalid("Unknown command type: " + type);
      }
    } catch (Exception e) {
      return new ParsedCommand.Invalid("Parse error: " + e.getMessage());
    }
  }

  // === Simple JSON utilities ===

  /** Escape and quote a string for JSON output. */
  private static String jsonString(String s) {
    if (s == null) return "null";
    StringBuilder sb = new StringBuilder(s.length() + 2);
    sb.append('"');
    for (int i = 0; i < s.length(); i++) {
      char c = s.charAt(i);
      switch (c) {
        case '"':
          sb.append("\\\"");
          break;
        case '\\':
          sb.append("\\\\");
          break;
        case '\b':
          sb.append("\\b");
          break;
        case '\f':
          sb.append("\\f");
          break;
        case '\n':
          sb.append("\\n");
          break;
        case '\r':
          sb.append("\\r");
          break;
        case '\t':
          sb.append("\\t");
          break;
        default:
          if (c < 0x20) {
            sb.append(String.format("\\u%04x", (int) c));
          } else {
            sb.append(c);
          }
      }
    }
    sb.append('"');
    return sb.toString();
  }

  /** Extract a string field value from JSON (simple parsing). */
  private static String extractStringField(String json, String fieldName) {
    String pattern = "\"" + fieldName + "\"";
    int idx = json.indexOf(pattern);
    if (idx < 0) return null;

    idx += pattern.length();
    // Skip whitespace and colon
    while (idx < json.length() && (json.charAt(idx) == ' ' || json.charAt(idx) == ':')) {
      idx++;
    }
    if (idx >= json.length() || json.charAt(idx) != '"') return null;

    idx++; // Skip opening quote
    StringBuilder sb = new StringBuilder();
    while (idx < json.length()) {
      char c = json.charAt(idx);
      if (c == '"') break;
      if (c == '\\' && idx + 1 < json.length()) {
        idx++;
        char escaped = json.charAt(idx);
        switch (escaped) {
          case '"':
            sb.append('"');
            break;
          case '\\':
            sb.append('\\');
            break;
          case 'n':
            sb.append('\n');
            break;
          case 'r':
            sb.append('\r');
            break;
          case 't':
            sb.append('\t');
            break;
          case 'b':
            sb.append('\b');
            break;
          case 'f':
            sb.append('\f');
            break;
          case 'u':
            if (idx + 4 < json.length()) {
              String hex = json.substring(idx + 1, idx + 5);
              sb.append((char) Integer.parseInt(hex, 16));
              idx += 4;
            }
            break;
          default:
            sb.append(escaped);
        }
      } else {
        sb.append(c);
      }
      idx++;
    }
    return sb.toString();
  }

  /** Extract a string array field value from JSON (simple parsing). */
  private static List<String> extractStringArray(String json, String fieldName) {
    String pattern = "\"" + fieldName + "\"";
    int idx = json.indexOf(pattern);
    if (idx < 0) return null;

    idx += pattern.length();
    // Skip to opening bracket
    while (idx < json.length() && json.charAt(idx) != '[') {
      idx++;
    }
    if (idx >= json.length()) return null;
    idx++; // Skip [

    List<String> result = new ArrayList<>();
    StringBuilder current = null;
    boolean inString = false;
    boolean escape = false;

    while (idx < json.length()) {
      char c = json.charAt(idx);

      if (escape) {
        if (current != null) {
          switch (c) {
            case '"':
              current.append('"');
              break;
            case '\\':
              current.append('\\');
              break;
            case 'n':
              current.append('\n');
              break;
            case 'r':
              current.append('\r');
              break;
            case 't':
              current.append('\t');
              break;
            default:
              current.append(c);
          }
        }
        escape = false;
      } else if (c == '\\' && inString) {
        escape = true;
      } else if (c == '"') {
        if (inString) {
          result.add(current.toString());
          current = null;
        } else {
          current = new StringBuilder();
        }
        inString = !inString;
      } else if (c == ']' && !inString) {
        break;
      } else if (inString && current != null) {
        current.append(c);
      }
      idx++;
    }

    return result;
  }

  // === Command encoding (parent sends these) ===

  public static String encodeRunSuite(String className, String framework, List<String> args) {
    StringBuilder sb = new StringBuilder();
    sb.append("{\"type\":\"RunSuite\",\"data\":{");
    sb.append("\"className\":").append(jsonString(className));
    sb.append(",\"framework\":").append(jsonString(framework));
    sb.append(",\"args\":[");
    for (int i = 0; i < args.size(); i++) {
      if (i > 0) sb.append(",");
      sb.append(jsonString(args.get(i)));
    }
    sb.append("]}}");
    return sb.toString();
  }

  public static String encodeShutdown() {
    return "{\"type\":\"Shutdown\"}";
  }

  // === Parsed command types ===

  public abstract static class ParsedCommand {
    public static final class RunSuite extends ParsedCommand {
      public final String className;
      public final String framework;
      public final List<String> args;

      public RunSuite(String className, String framework, List<String> args) {
        this.className = className;
        this.framework = framework;
        this.args = args;
      }
    }

    public static final class Shutdown extends ParsedCommand {}

    public static final class GetThreadDump extends ParsedCommand {}

    public static final class Invalid extends ParsedCommand {
      public final String message;

      public Invalid(String message) {
        this.message = message;
      }
    }
  }
}

File: bleep-core/src/scala/bleep/testing/TestProtocol.scala
package bleep.testing

import bleep.bsp.protocol.SuiteOutcome
import io.circe._
import io.circe.generic.semiauto._
import io.circe.syntax._

/** Protocol messages for communication between bleep and forked test JVMs.
  *
  * Uses simple JSON-over-stdin/stdout for portability and debuggability.
  */
object TestProtocol {

  // === Commands (parent -> forked JVM) ===

  /** Commands sent from bleep to the forked test runner */
  sealed trait TestCommand

  object TestCommand {

    /** Run a test suite.
      *
      * `selection` carries the execution decision, not just a name — see [[FrameworkSelection]] for why the fork is told rather than asked to work it out.
      */
    case class RunSuite(
        className: String,
        selection: FrameworkSelection,
        args: List[String]
    ) extends TestCommand

    /** Run a whole project's JUnit-Platform suites in ONE launcher execution, at a bleep-chosen degree of parallelism.
      *
      * This is the maven-surefire shape for JUnit: all of a module's classes go through one `launcher.execute()`, so anything scoped to that one execution — an
      * application booted for the run, a shared `LauncherSessionListener` registration — is set up once and reused across every class, instead of rebuilt per
      * class. `parallelism` is bleep's decision, not junit's: 1 serialises (what a singleton-per-JVM application needs), N runs N classes at once. junit's
      * engine is the executor of that number, nothing more. Only the JUnit-Platform runner honours this; sbt test-interface frameworks are driven
      * suite-by-suite on bleep's own threads, where there is no cross-suite execution scope to preserve.
      */
    case class RunSuites(
        classNames: List[String],
        parallelism: Int,
        selection: FrameworkSelection,
        args: List[String]
    ) extends TestCommand

    /** Cancel one in-flight suite by class name without touching the fork or its siblings.
      *
      * Only meaningful when the fork is running several suites at once (a per-project shared session): it interrupts just that suite's thread, so the other
      * suites sharing the JVM keep running. An exclusive fork cancels by having its socket closed instead — there is only the one suite, and the whole process
      * goes. See [[bleep.model.TestJvmMode]].
      */
    case class CancelSuite(className: String) extends TestCommand

    /** Gracefully shut down the forked JVM */
    case object Shutdown extends TestCommand

    /** Get a thread dump from the forked JVM */
    case object GetThreadDump extends TestCommand

    /** The wire strings, read from the Java enum the fork parses with, so there is exactly one definition of them. */
    private object RunnerWire {
      val JUnitPlatform: String = runner.TestProtocol.RunnerKind.JUNIT_PLATFORM.wire()
      val SbtTestInterface: String = runner.TestProtocol.RunnerKind.SBT_TEST_INTERFACE.wire()
    }

    /** Hand-written rather than derived, because the wire form is flat while the model is a sum: `runner` names the mechanism and `frameworkClass` is present
      * exactly when that mechanism needs one. The Java parser in the fork enforces the same pairing, and both sides reject a mismatch rather than defaulting.
      */
    implicit val runSuiteEncoder: Encoder[RunSuite] = Encoder.instance { rs =>
      val (runner, frameworkClass) = rs.selection match {
        case FrameworkSelection.JUnitPlatform(_)         => (RunnerWire.JUnitPlatform, None)
        case FrameworkSelection.SbtTestInterface(_, cls) => (RunnerWire.SbtTestInterface, Some(cls))
        case FrameworkSelection.PlatformRunner(name)     =>
          // Unreachable by construction: the test handler routes JS/Native suites to their own runners before any fork exists. If it ever is reached, the
          // fork cannot run this suite and a placeholder command would only move the failure somewhere less obvious.
          sys.error(s"$name is run by its platform's own runner, not by a forked JVM — it must never be sent over TestProtocol (suite ${rs.className})")
      }
      Json.obj(
        "className" -> rs.className.asJson,
        "framework" -> rs.selection.displayName.asJson,
        "runner" -> runner.asJson,
        "frameworkClass" -> frameworkClass.asJson,
        "args" -> rs.args.asJson
      )
    }

    implicit val runSuiteDecoder: Decoder[RunSuite] = Decoder.instance { cursor =>
      for {
        className <- cursor.downField("className").as[String]
        displayName <- cursor.downField("framework").as[String]
        runner <- cursor.downField("runner").as[String]
        frameworkClass <- cursor.downField("frameworkClass").as[Option[String]]
        args <- cursor.downField("args").as[List[String]]
        selection <- (runner, frameworkClass) match {
          case (RunnerWire.JUnitPlatform, _)            => Right(FrameworkSelection.JUnitPlatform(displayName))
          case (RunnerWire.SbtTestInterface, Some(cls)) => Right(FrameworkSelection.SbtTestInterface(displayName, cls))
          case (RunnerWire.SbtTestInterface, None)      => Left(DecodingFailure(s"${RunnerWire.SbtTestInterface} requires frameworkClass", cursor.history))
          case (other, _)                               => Left(DecodingFailure(s"Unknown runner: $other", cursor.history))
        }
      } yield RunSuite(className, selection, args)
    }

    // Only JUnit-Platform suites are ever batched into one execution (that is the runner with a cross-class scope worth preserving), so the wire form carries
    // the one runner + display name shared by all the classes, plus the class list and bleep's parallelism.
    implicit val runSuitesEncoder: Encoder[RunSuites] = Encoder.instance { rs =>
      // A batch's classes all share one framework, so one runner + (for sbt) one frameworkClass covers them. JUnit Platform needs no frameworkClass; sbt does.
      val (runner, frameworkClass) = rs.selection match {
        case FrameworkSelection.JUnitPlatform(_)         => (RunnerWire.JUnitPlatform, None)
        case FrameworkSelection.SbtTestInterface(_, cls) => (RunnerWire.SbtTestInterface, Some(cls))
        case other => sys.error(s"RunSuites is for JUnit-Platform or sbt-test-interface; got $other for ${rs.classNames.mkString(", ")}")
      }
      Json.obj(
        "classNames" -> rs.classNames.asJson,
        "parallelism" -> rs.parallelism.asJson,
        "framework" -> rs.selection.displayName.asJson,
        "runner" -> runner.asJson,
        "frameworkClass" -> frameworkClass.asJson,
        "args" -> rs.args.asJson
      )
    }

    implicit val runSuitesDecoder: Decoder[RunSuites] = Decoder.instance { cursor =>
      for {
        classNames <- cursor.downField("classNames").as[List[String]]
        parallelism <- cursor.downField("parallelism").as[Int]
        displayName <- cursor.downField("framework").as[String]
        runner <- cursor.downField("runner").as[String]
        frameworkClass <- cursor.downField("frameworkClass").as[Option[String]]
        args <- cursor.downField("args").as[List[String]]
        selection <- (runner, frameworkClass) match {
          case (RunnerWire.JUnitPlatform, _)            => Right(FrameworkSelection.JUnitPlatform(displayName))
          case (RunnerWire.SbtTestInterface, Some(cls)) => Right(FrameworkSelection.SbtTestInterface(displayName, cls))
          case (RunnerWire.SbtTestInterface, None) => Left(DecodingFailure(s"${RunnerWire.SbtTestInterface} RunSuites requires frameworkClass", cursor.history))
          case (other, _)                          => Left(DecodingFailure(s"Unknown runner for RunSuites: $other", cursor.history))
        }
      } yield RunSuites(classNames, parallelism, selection, args)
    }

    implicit val encoder: Encoder[TestCommand] = Encoder.instance {
      case rs: RunSuite    => Json.obj("type" -> "RunSuite".asJson, "data" -> rs.asJson)
      case rs: RunSuites   => Json.obj("type" -> "RunSuites".asJson, "data" -> rs.asJson)
      case cs: CancelSuite => Json.obj("type" -> "CancelSuite".asJson, "className" -> cs.className.asJson)
      case Shutdown        => Json.obj("type" -> "Shutdown".asJson)
      case GetThreadDump   => Json.obj("type" -> "GetThreadDump".asJson)
    }

    implicit val decoder: Decoder[TestCommand] = Decoder.instance { cursor =>
      cursor.downField("type").as[String].flatMap {
        case "RunSuite"      => cursor.downField("data").as[RunSuite]
        case "RunSuites"     => cursor.downField("data").as[RunSuites]
        case "CancelSuite"   => cursor.downField("className").as[String].map(CancelSuite(_))
        case "Shutdown"      => Right(Shutdown)
        case "GetThreadDump" => Right(GetThreadDump)
        case other           => Left(DecodingFailure(s"Unknown command type: $other", cursor.history))
      }
    }
  }

  // === Responses (forked JVM -> parent) ===

  /** Responses sent from the forked test runner back to bleep */
  sealed trait TestResponse

  object TestResponse {

    /** Test runner is ready to receive commands */
    case object Ready extends TestResponse

    /** A test has started */
    case class TestStarted(
        suite: String,
        test: String
    ) extends TestResponse

    /** A test has finished */
    case class TestFinished(
        suite: String,
        test: String,
        status: String, // passed, failed, error, skipped, ignored, cancelled, pending
        durationMs: Long,
        message: Option[String],
        throwable: Option[String],
        // Where in the suite the failure was raised, when the forked runner could recover it from the throwable.
        // Absent for passing tests, for failures thrown outside the suite class, and from runners that do not report it.
        location: Option[bleep.bsp.protocol.BleepBspProtocol.SourceLocation]
    ) extends TestResponse

    /** A test suite has completed. `outcome` is reconstructed from the flat wire fields (the Java forked runner emits a `kind` discriminator plus counts) into
      * the [[SuiteOutcome]] ADT so nothing downstream re-derives meaning from an all-zero count tuple.
      */
    case class SuiteDone(
        suite: String,
        outcome: SuiteOutcome,
        durationMs: Long
    ) extends TestResponse

    /** Log output from test */
    case class Log(
        level: String,
        message: String,
        suite: Option[String]
    ) extends TestResponse

    /** An error occurred in the test runner itself */
    case class Error(
        message: String,
        throwable: Option[String]
    ) extends TestResponse

    /** The single batched execution (a `RunSuites`) has returned. Every class's own terminal (`SuiteDone`/error) has already been sent and demultiplexed to its
      * suite; this just tells the parent the one execute is over, so it can stop reading without counting terminals against a class list.
      */
    case object BatchComplete extends TestResponse

    /** Thread dump from the forked JVM */
    case class ThreadDump(
        threads: List[ThreadInfo]
    ) extends TestResponse

    /** Information about a single thread */
    case class ThreadInfo(
        name: String,
        state: String,
        stackTrace: List[String]
    )

    implicit val threadInfoEncoder: Encoder[ThreadInfo] = deriveEncoder
    implicit val threadInfoDecoder: Decoder[ThreadInfo] = deriveDecoder

    implicit val threadDumpEncoder: Encoder[ThreadDump] = deriveEncoder
    implicit val threadDumpDecoder: Decoder[ThreadDump] = deriveDecoder

    implicit val testStartedEncoder: Encoder[TestStarted] = deriveEncoder
    implicit val testStartedDecoder: Decoder[TestStarted] = deriveDecoder

    implicit val testFinishedEncoder: Encoder[TestFinished] = deriveEncoder
    implicit val testFinishedDecoder: Decoder[TestFinished] = deriveDecoder

    // The wire is flat (kind discriminator + counts + optional message/throwable), matching the
    // hand-rolled JSON the Java forked runner emits; the outcome ADT is (re)constructed here.
    implicit val suiteDoneEncoder: Encoder[SuiteDone] = Encoder.instance { sd =>
      val base = Json.obj(
        "suite" -> sd.suite.asJson,
        "outcome" -> SuiteOutcome.tagOf(sd.outcome).asJson,
        "passed" -> sd.outcome.passedCount.asJson,
        "failed" -> sd.outcome.failedCount.asJson,
        "skipped" -> sd.outcome.skippedCount.asJson,
        "ignored" -> sd.outcome.ignoredCount.asJson,
        "durationMs" -> sd.durationMs.asJson
      )
      sd.outcome match {
        case SuiteOutcome.Errored(message, throwable) =>
          base.deepMerge(Json.obj("message" -> message.asJson, "throwable" -> throwable.asJson))
        case _ => base
      }
    }
    implicit val suiteDoneDecoder: Decoder[SuiteDone] = Decoder.instance { c =>
      for {
        suite <- c.downField("suite").as[String]
        // `outcome` is optional: a forked test-runner jar that predates the SuiteOutcome ADT (a
        // version-skewed test classpath — e.g. an older bleep-test-runner resolved from the build)
        // emits only the flat counts. Reconstruct the outcome from them rather than failing to
        // decode. This is the forked-runner analog of the version-tolerant BSP SuiteFinished codec.
        kind <- c.downField("outcome").as[Option[String]]
        passed <- c.getOrElse("passed")(0)
        failed <- c.getOrElse("failed")(0)
        skipped <- c.getOrElse("skipped")(0)
        ignored <- c.getOrElse("ignored")(0)
        durationMs <- c.downField("durationMs").as[Long]
        message <- c.downField("message").as[Option[String]]
        throwable <- c.downField("throwable").as[Option[String]]
      } yield {
        // `fromWire` already routes a tag it doesn't recognise through the same `degrade` policy the
        // BSP SuiteFinished codec uses, so both boundaries treat version skew identically.
        val outcome = kind match {
          case Some(k) => SuiteOutcome.fromWire(k, passed, failed, skipped, ignored, message, throwable)
          case None    => SuiteOutcome.degrade(None, passed, failed, skipped, ignored)
        }
        SuiteDone(suite, outcome, durationMs)
      }
    }

    implicit val logEncoder: Encoder[Log] = deriveEncoder
    implicit val logDecoder: Decoder[Log] = deriveDecoder

    implicit val errorEncoder: Encoder[Error] = deriveEncoder
    implicit val errorDecoder: Decoder[Error] = deriveDecoder

    implicit val encoder: Encoder[TestResponse] = Encoder.instance {
      case Ready            => Json.obj("type" -> "Ready".asJson)
      case ts: TestStarted  => Json.obj("type" -> "TestStarted".asJson, "data" -> ts.asJson)
      case tf: TestFinished => Json.obj("type" -> "TestFinished".asJson, "data" -> tf.asJson)
      case sd: SuiteDone    => Json.obj("type" -> "SuiteDone".asJson, "data" -> sd.asJson)
      case l: Log           => Json.obj("type" -> "Log".asJson, "data" -> l.asJson)
      case e: Error         => Json.obj("type" -> "Error".asJson, "data" -> e.asJson)
      case td: ThreadDump   => Json.obj("type" -> "ThreadDump".asJson, "data" -> td.asJson)
      case BatchComplete    => Json.obj("type" -> "BatchComplete".asJson)
    }

    implicit val decoder: Decoder[TestResponse] = Decoder.instance { cursor =>
      cursor.downField("type").as[String].flatMap {
        case "Ready"         => Right(Ready)
        case "TestStarted"   => cursor.downField("data").as[TestStarted]
        case "TestFinished"  => cursor.downField("data").as[TestFinished]
        case "SuiteDone"     => cursor.downField("data").as[SuiteDone]
        case "Log"           => cursor.downField("data").as[Log]
        case "Error"         => cursor.downField("data").as[Error]
        case "ThreadDump"    => cursor.downField("data").as[ThreadDump]
        case "BatchComplete" => Right(BatchComplete)
        case other           => Left(DecodingFailure(s"Unknown response type: $other", cursor.history))
      }
    }
  }

  // === Encoding/Decoding utilities ===

  /** Encode a command to a single line of JSON */
  def encodeCommand(cmd: TestCommand): String =
    cmd.asJson.noSpaces

  /** Decode a command from a JSON line */
  def decodeCommand(line: String): Either[io.circe.Error, TestCommand] =
    io.circe.parser.decode[TestCommand](line)

  /** Encode a response to a single line of JSON */
  def encodeResponse(resp: TestResponse): String =
    resp.asJson.noSpaces

  /** Decode a response from a JSON line */
  def decodeResponse(line: String): Either[io.circe.Error, TestResponse] =
    io.circe.parser.decode[TestResponse](line)
}
